pkgs/development/haskell-modules/configuration-common.nix

```diff
@@ -1593,7 +1585,7 @@ self: super: {
 
   # 2020-12-05: http-client is fixed on too old version
   essence-of-live-coding-warp = doJailbreak (super.essence-of-live-coding-warp.override {
-    http-client = self.http-client_0_7_8;
+    http-client = self.http-client_0_7_9;
```

Comment:
I'm not sure I understand what is going on here.  Why is `http-client` going from version `0.7.8` to `0.7.9`?

I was under the impression that we generally don't want to update versions of things on the release branches, unless there was some reason to explicitly do it.

In this same file, above we have `haskeline_0_8_1_2` becoming `haskeline_0_8_2`, but I guess this means that`haskeline_0_8_1_2` is no longer available?  If so, I imagine this approach won't work as-is, since we definitely don't want to remove derivations that people may be using.  Something like this is of course okay to do on `master` / unstable, but not for the release branches.

Reply:
This change does not consider the versioned attributes part of the stable interface of the release branch. This is fair I think: They are mostly not well tested (most probably don't compile) and intended for internal use. If you need a certain version of a package, it is always better to use `callHackage`.

The upgrades done in the change are mostly to fix eval errors, alternatively, I could've added an `extra-packages` directive. However, I don't want to inflate `hackage-packages.nix` unnecessarily and we are allowed to do non-breaking (as per PVP) package updates.

Since this would need to go through a staging cycle anyways, we also have the opportunity to catch edge cases and regressions from this.

Preserving all versioned attributes is unfeasible: This would mean that we'd have to retain a versioned attribute for every version of the package released during the lifetime of the release branch in order to never remove an attribute someone *may* be using.

Reply:
We could also consider doing this only for 21.11 and clearly communicating this change in attitude in the release notes.

Reply:
> This change does not consider the versioned attributes part of the stable interface of the release branch. This is fair I think: They are mostly not well tested (most probably don't compile) and intended for internal use. 

I don't think this is a reasonable change.  Nowhere in our documentation that suggests that the versioned Haskell packages are internal or not intended for end users.  My communication with users has always suggested that the versioned attributes are completely fine to use.

There is no way this change is okay to make on a release branch.

> If you need a certain version of a package, it is always better to use callHackage.

Some of the versioned packages do have specific fixes in files like `configuration-common.nix`.  This sometimes makes them easier to use than `callHackage`.

--------------------------------------------------------

If we did want to make the versioned packages internal and have people stop using them, the following would be nice to do:

- Go through a release cycle where we have a big warning in the release notes.  
- Possibly put some sort of `builtins.trace` log message on all the versioned packages that aren't in `extra-packages` saying they are not meant to be used.
- Possibly renamed the versioned packaged to something like `__lens_5_0_0` to better indicate that they are internal.
- Possibly just stop generating the versioned packages that aren't in `extra-packages`?  There are only a few tens of lines using versioned packages in `configuration-common.nix` and `configuration-nix.nix`, so adding these by hand shouldn't be too crazy.

Reply:
> Possibly put some sort of builtins.trace log message on all the versioned packages that aren't in extra-packages saying they are not meant to be used.

This wouldn't work since it'd break our use of versioned attributes.

> Although, I don't know how I feel about going forward with considering the versioned packages as internal. Sometimes they are convenient to have.

We don't need to forbid their use, they just change. It's a personal trade-off to make: Do I want to fix this (possibly) every few weeks and try if I can use an even newer version?

In general, maybe we should take a step back and talk about stable maintenance more generally? Seems like we are different levels of conservative here. I can offer to draw up a RFC of what I have in mind.

Reply:
> This wouldn't work since it'd break our use of versioned attributes.

We would have to start using the versioned packages differently.  For instance, we'd have to manually add a package to `extra-packages` every time we wanted to use a versioned attribute.

> maybe we should take a step back and talk about stable maintenance more generally? Seems like we are different levels of conservative here.

Yeah, that sounds like a good idea.

We should probably also discuss the general NixOS rules/guarantees for release branches.  Here's my understanding:

- It is never okay to remove derivations/attrs except in exceptional circumstances (like security problems, or non-open-source software that no longer works).
- Major version bumps are never okay, except in exceptional circumstances (like the previous rule).
- Backporting new versions of software at new attr names is generally okay.
- Minor version bumps for bug fixes are okay (although a patch to the current version would be preferred).
- Minor version bumps for new features (not bug fixes) are generally discouraged, although sometimes they happen depending on the maintainer.

Although I don't know if I've ever seen these rules explicitly written down somewhere.  Do you know if these rules are written down anywhere?

Reply:
> We would have to start using the versioned packages differently. For instance, we'd have to manually add a package to extra-packages every time we wanted to use a versioned attribute.

I consider it a feature that an eval error is introduced as soon as the versioned attribute is not the latest anymore.

> Although I don't know if I've ever seen these rules explicitly written down somewhere. Do you know if these rules are written down anywhere?

<https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#backporting-changes>

Reply:
> I consider it a feature that an eval error is introduced as soon as the versioned attribute is not the latest anymore.

I don't necessarily feel strongly either way. But it does create more work for us, since we always have to fix up these eval errors before merging `haskell-updates` to `master`.  If we instead just added everything we need to `extra-packages`, then we normally wouldn't have to worry about eval errors.  Although this would come at the expense of likely having unused packages accumulate in `extra-packages` over time.

> https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#backporting-changes

Thanks for this link.

This seems to roughly coincidence with my last message, except that:

- it doesn't mention anything about removing derivations, or what could be considered an internal-use-only attribute
- it doesn't seem like minor version bumps are discouraged (as long as they are non-breaking for current users).

-------------------------------

So it seems like the things we need to hash out are the following.  I guess this would be a good starting point for an RFC:

- Are the versioned Haskell packages are meant for internal-use-only?  Should end users refrain from using them?  Is this documented somewhere?  If this isn't documented anywhere, then I don't we can remove/updated versioned Haskell packages from an existing release branch.  It may be possible for a future release as long as we document it clearly in the release notes.
- If the versioned Haskell packages are meant for internal-user-only, then is there some way we can make that more obvious for end-users?  (For instance, changing the names to something like `__lens_0_5_0_0` with two leading underscores.)
- Is there a good reason to even have the versioned attributes generated automatically?  Could we move to manually adding things to `extra-packages`?
- Could we just freeze the versioned packages as well?

----------------------------------

I don't know if I made it clear above, but I am in favor with this approach in general.  I'm just worried about breaking the guarantees that come with the Nixpkgs _release_ branches.